File: packages/venrescate-core-ts/test/vector-test.ts
// venrescate-core-ts: cross-language vector test
// Loads test-vectors/*.json and asserts exact agreement with gen-vectors.js outputs.
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { canon, keyFromSeed, sign, verify, fold, IDS } from '../src/index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const VECTORS = join(__dirname, '..', '..', '..', 'test-vectors');

function load(name: string) {
  return JSON.parse(readFileSync(join(VECTORS, name), 'utf8'));
}

const keys = load('keys.json');
const cryptoVectors = load('crypto-vectors.json');
const foldVectors = load('fold-vectors.json');
const clockVectors = load('clock-vectors.json');

let passed = 0;
let failed = 0;

function assert(cond: boolean, msg: string) {
  if (cond) { passed++; return true; }
  console.error(`FAIL: ${msg}`);
  failed++;
  return false;
}

// 1. Derive identities from keys.json seeds
console.log('\n=== Identity derivation ===');
const identities: Record<string, ReturnType<typeof keyFromSeed>> = {};
for (const [name, k] of Object.entries(keys) as [string, any][]) {
  const id = keyFromSeed(k.seed_hex);
  identities[name] = id;
  assert(id.publicKeyB64u === k.public_key_b64u, `${name}: public key matches`);
  assert(id.deviceId === k.device_id, `${name}: device_id matches`);
}

const pubRawById: Record<string, Uint8Array> = {};
for (const [, id] of Object.entries(identities)) {
  pubRawById[id.deviceId] = id.publicKeyRaw;
}

// 2. Crypto vectors: sign & verify
console.log('\n=== Crypto vectors ===');
assert(cryptoVectors.cases[0].expect_verify === true, 'valid_signed_sos: expect true');
const sosAlice = cryptoVectors.cases[0].record;

// Re-sign: remove sig + content_hash, canonicalize, sign with alice
const { sig, content_hash, ...recordBody } = sosAlice;
const reSigned = sign(recordBody, identities.alice);
assert(reSigned.contentHash === (sosAlice as any).content_hash, 'valid_signed_sos: content_hash matches');
assert(reSigned.sig === (sosAlice as any).sig, 'valid_signed_sos: sig matches');

// Canonical bytes match
assert(canon(recordBody) === cryptoVectors.cases[0].canonical_bytes_utf8, 'valid_signed_sos: canonical bytes match');

// Verify
const verifyResult = verify(sosAlice, identities.alice.publicKeyRaw);
assert(verifyResult === true, 'valid_signed_sos: verify returns true');

// Tampered
const tampered = cryptoVectors.cases[1].record;
const tamperedVerify = verify(tampered, identities.alice.publicKeyRaw);
assert(tamperedVerify === false, 'tampered_payload: verify returns false');

// Wrong key
const wrongKeyVerify = verify(sosAlice, identities.bob.publicKeyRaw);
assert(wrongKeyVerify === false, 'wrong_key: verify returns false');

// 3. Fold vectors
console.log('\n=== Fold vectors ===');
for (const scenario of foldVectors.scenarios) {
  const name = scenario.name;
  const result = fold(scenario.report, scenario.attestations, pubRawById, scenario.subject_device_id || null);

  const expected = scenario.expected;
  assert(result.tier === expected.tier, `${name}: tier ${result.tier} === ${expected.tier}`);
  assert(result.tierName === expected.tierName, `${name}: tierName ${result.tierName} === ${expected.tierName}`);
  assert(result.verified === expected.verified, `${name}: verified ${result.verified} === ${expected.verified}`);
  assert(result.disputed === expected.disputed, `${name}: disputed ${result.disputed} === ${expected.disputed}`);
  assert(result.locationVerified === expected.location_verified, `${name}: location_verified ${result.locationVerified} === ${expected.location_verified}`);

  if (expected.note && result.note !== expected.note) {
    console.error(`  note mismatch: expected "${expected.note}", got "${result.note}"`);
  }
}

// 4. Clock vectors (logic check)
console.log('\n=== Clock vectors ===');
assert(clockVectors.cases[0].expected_order[0] === 'ELpoLIrRNROXHotW:1', 'causal_order: first record');
assert(clockVectors.cases[0].expected_order[1] === 'ELpoLIrRNROXHotW:2', 'causal_order: second record');
assert(clockVectors.cases[1].prio === 0, 'p0_future: prio is 0');
assert(clockVectors.cases[1].expected.includes('ACCEPTED'), 'p0_future: expected ACCEPTED');

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exitCode = 1;
else console.log('ALL VECTORS PASS — TypeScript');

File: packages/venrescate-core/src/test/kotlin/venrescate/VectorTest.kt
package venrescate

import venrescate.crypto.VenRescateIdentity
import venrescate.crypto.Crypto
import venrescate.trust.VerificationFold
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File

object VectorTest {
    private val gson = Gson()
    private val vectorsDir = File(System.getProperty("user.dir"), "test-vectors")

    private fun load(name: String): Map<String, Any?> {
        val file = File(vectorsDir, name)
        val type = object : TypeToken<Map<String, Any?>>() {}.type
        return gson.fromJson(file.readText(), type)
    }

    private var passed = 0
    private var failed = 0

    private fun assert(cond: Boolean, msg: String) {
        if (cond) { passed++ }
        else { println("FAIL: $msg"); failed++ }
    }

    @JvmStatic
    fun main(args: Array<String>) {
        // Load vectors
        val keys = load("keys.json")
        val cryptoVectors = load("crypto-vectors.json")
        val foldVectors = load("fold-vectors.json")
        val clockVectors = load("clock-vectors.json")

        // 1. Derive identities
        println("\n=== Identity derivation ===")
        val identities = mutableMapOf<String, VenRescateIdentity>()
        val pubRawById = mutableMapOf<String, ByteArray>()

        @Suppress("UNCHECKED_CAST")
        val keysMap = keys as Map<String, Map<String, String>>
        for ((name, k) in keysMap) {
            val id = Crypto.keyFromSeed(k["seed_hex"]!!)
            identities[name] = id
            assert(id.publicKeyB64u == k["public_key_b64u"], "$name: public key matches")
            assert(id.deviceId == k["device_id"], "$name: device_id matches")
            pubRawById[id.deviceId] = id.publicKeyRaw
        }

        // 2. Crypto vectors
        println("\n=== Crypto vectors ===")
        @Suppress("UNCHECKED_CAST")
        val cases = cryptoVectors["cases"] as List<Map<String, Any?>>

        // valid_signed_sos
        val case0 = cases[0]
        @Suppress("UNCHECKED_CAST")
        val sosAlice = case0["record"] as Map<String, Any?>
        assert(case0["expect_verify"] == true, "valid_signed_sos: expect true")

        // Re-sign
        val recordBody = sosAlice.filterKeys { it != "sig" && it != "content_hash" }
        val reSigned = Crypto.sign(recordBody, identities["alice"]!!)
        assert(reSigned.contentHash == sosAlice["content_hash"], "valid_signed_sos: content_hash matches")
        assert(reSigned.sig == sosAlice["sig"], "valid_signed_sos: sig matches")

        // Canonical bytes
        assert(reSigned.canonical == case0["canonical_bytes_utf8"], "valid_signed_sos: canonical bytes match")

        // Verify
        assert(Crypto.verify(sosAlice, identities["alice"]!!.publicKeyRaw), "valid_signed_sos: verify returns true")

        // Tampered
        @Suppress("UNCHECKED_CAST")
        val tampered = cases[1]["record"] as Map<String, Any?>
        assert(!Crypto.verify(tampered, identities["alice"]!!.publicKeyRaw), "tampered_payload: verify returns false")

        // Wrong key
        assert(!Crypto.verify(sosAlice, identities["bob"]!!.publicKeyRaw), "wrong_key: verify returns false")

        // 3. Fold vectors
        println("\n=== Fold vectors ===")
        @Suppress("UNCHECKED_CAST")
        val scenarios = foldVectors["scenarios"] as List<Map<String, Any?>>
        for (scenario in scenarios) {
            val name = scenario["name"] as String
            @Suppress("UNCHECKED_CAST")
            val report = scenario["report"] as Map<String, Any?>
            @Suppress("UNCHECKED_CAST")
            val attestations = scenario["attestations"] as List<Map<String, Any?>>
            val subjectId = scenario["subject_device_id"] as? String

            val result = VerificationFold.fold(report, attestations, pubRawById, subjectId)
            @Suppress("UNCHECKED_CAST")
            val expected = scenario["expected"] as Map<String, Any?>

            assert(result.tier == (expected["tier"] as Double).toInt(), "$name: tier ${result.tier} == ${(expected["tier"] as Double).toInt()}")
            assert(result.tierName == expected["tierName"], "$name: tierName ${result.tierName} == ${expected["tierName"]}")
            assert(result.verified == expected["verified"], "$name: verified")
            assert(result.disputed == expected["disputed"], "$name: disputed")
            assert(result.locationVerified == expected["location_verified"], "$name: location_verified")
        }

        // 4. Clock vectors
        println("\n=== Clock vectors ===")
        @Suppress("UNCHECKED_CAST")
        val clockCases = clockVectors["cases"] as List<Map<String, Any?>>
        @Suppress("UNCHECKED_CAST")
        val expectedOrder = clockCases[0]["expected_order"] as List<String>
        assert(expectedOrder[0] == "ELpoLIrRNROXHotW:1", "causal_order: first record")
        assert(expectedOrder[1] == "ELpoLIrRNROXHotW:2", "causal_order: second record")
        assert((clockCases[1]["prio"] as Double).toInt() == 0, "p0_future: prio is 0")
        assert((clockCases[1]["expected"] as String).contains("ACCEPTED"), "p0_future: expected ACCEPTED")

        println("\n$passed passed, $failed failed")
        if (failed > 0) throw AssertionError("$failed vector assertions failed")
        else println("ALL VECTORS PASS — Kotlin")
    }
}
